[One Workflow] Data transform: regex extract and replace

talboren opened the pull request.

talboren committed:
Add data regex extraction and replacement steps

This commit introduces two new workflow step definitions: `data.regex_extract` and `data.regex_replace`. These steps are now exported from the data module and registered in both the server and public step registries. Additionally, the approved step definitions have been updated to include these new steps.

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
Merge branch 'main' into 14805-transform-regex-extract-and-replace

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376455)
* Commit: 7bb0a11070ab3fa2b13e3ed073fcbf05a17e8be9

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/376455#019b4b30-c569-4285-8d31-770ae3bbc5ff)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376455","number":376455,"commit":"7bb0a11070ab3fa2b13e3ed073fcbf05a17e8be9"}],"number":376455}
buildkite-pr-comment-->

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
fix

talboren committed:
Merge branch '14805-transform-regex-extract-and-replace' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376480)
* Commit: 3b632c193cc8400110e570e2c475c5773cabd40b

### Failed CI Steps
* [Scout: [ platform / workflows_extensions ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b86-3356-4d14-be14-4987b25ed2db)
* [Scout: [ platform / workflows_extensions ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b92-d240-40d0-8bb6-6ea892b5f7ca)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b86-3356-4d14-be14-4987b25ed2db) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b86-3356-4d14-be14-4987b25ed2db/artifacts/019b4b92-b2c1-4429-a79e-409bae373787) Scout: [ platform / workflows_extensions ] plugin / serverless-es - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b92-d240-40d0-8bb6-6ea892b5f7ca) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b92-d240-40d0-8bb6-6ea892b5f7ca/artifacts/019b4b9f-309d-41bd-abd5-d37d765bf271) Scout: [ platform / workflows_extensions ] plugin / serverless-es - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b86-3356-4d14-be14-4987b25ed2db) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b86-3356-4d14-be14-4987b25ed2db/artifacts/019b4b92-b2c1-4862-963b-6043ae9c5945) Scout: [ platform / workflows_extensions ] plugin / serverless-oblt - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b92-d240-40d0-8bb6-6ea892b5f7ca) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b92-d240-40d0-8bb6-6ea892b5f7ca/artifacts/019b4b9f-309c-4748-b9b8-05c2446f47fe) Scout: [ platform / workflows_extensions ] plugin / serverless-oblt - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b86-3356-4d14-be14-4987b25ed2db) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b86-3356-4d14-be14-4987b25ed2db/artifacts/019b4b92-b2c1-4f6e-97e7-999d31e19ecb) Scout: [ platform / workflows_extensions ] plugin / serverless-security - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b92-d240-40d0-8bb6-6ea892b5f7ca) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b92-d240-40d0-8bb6-6ea892b5f7ca/artifacts/019b4b9f-309b-43a8-9eb6-aa19ef34de14) Scout: [ platform / workflows_extensions ] plugin / serverless-security - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b86-3356-4d14-be14-4987b25ed2db) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b86-3356-4d14-be14-4987b25ed2db/artifacts/019b4b92-b2c2-4782-a904-5436e67ed2ae) Scout: [ platform / workflows_extensions ] plugin / should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b92-d240-40d0-8bb6-6ea892b5f7ca) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b92-d240-40d0-8bb6-6ea892b5f7ca/artifacts/019b4b9f-309d-4afd-8a55-c4ef6677d44c) Scout: [ platform / workflows_extensions ] plugin / should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b86-3356-4d14-be14-4987b25ed2db) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b86-3356-4d14-be14-4987b25ed2db/artifacts/019b4b92-b2c1-4f43-b8cf-ed4ea2ac22b4) Scout: [ platform / workflows_extensions ] plugin / stateful - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376480#019b4b92-d240-40d0-8bb6-6ea892b5f7ca) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376480/jobs/019b4b92-d240-40d0-8bb6-6ea892b5f7ca/artifacts/019b4b9f-309c-47a5-a23e-78fe11769c6f) Scout: [ platform / workflows_extensions ] plugin / stateful - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team

### History
* :broken_heart: [Build #376455](https://buildkite.com/elastic/kibana-pull-request/builds/376455) failed 7bb0a11070ab3fa2b13e3ed073fcbf05a17e8be9

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376480","number":376480,"commit":"3b632c193cc8400110e570e2c475c5773cabd40b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376455","number":376455,"commit":"7bb0a11070ab3fa2b13e3ed073fcbf05a17e8be9"}],"number":376480}
buildkite-pr-comment-->

h88 reviewed (commented):
Initial important questions that I see Copilot already touched as well:

- How do we prevent ReDoS attacks?
- Should we add execution timeouts / pattern validation?
- Should we limit pattern length/complexity or detect known ReDoS patterns?
- Should we enforce limits at the Zod schema level in addition to the handler?

TBC & verified locally after addressing / deciding on the initial CR...

talboren commented:
> Initial important questions that I see Copilot already touched as well:
> 
> * How do we prevent ReDoS attacks?
> * Should we add execution timeouts / pattern validation?
> * Should we limit pattern length/complexity or detect known ReDoS patterns?
> * Should we enforce limits at the Zod schema level in addition to the handler?
> 
> TBC & verified locally after addressing / deciding on the initial CR...

Addressing the different points by Copilot soon.

In the meantime, re: ReDos protection and timeout mechanism for regex execution

True regex timeouts aren't feasible in JavaScript's single-threaded runtime - we can't interrupt a running match() or replace() operation without worker threads (excessive overhead).

The 100KB input length limit I'm adding is the practical defense here, as most ReDoS attacks require long strings (10K+ chars) to trigger catastrophic backtracking. We can also check abortSignal between array iterations for graceful cancellation, though this can't stop an already-running regex operation.

LMK what do you think

talboren committed:
Refactor regex extract and replace steps for improved validation and performance

- Added input validation for source and length in both dataRegexExtract and dataRegexReplace steps.
- Enhanced regex handling by introducing utility functions for regex creation and validation.
- Updated documentation to include security notes regarding regex performance issues.
- Added comprehensive tests for global flag behavior in dataRegexExtract step and detailed logging in dataRegexReplace step.

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
update hashes

talboren committed:
Update regex handling in data steps for improved functionality

- Updated handler hashes in approved step definitions.
- Refactored dataRegexExtract and dataRegexReplace steps to enhance regex validation and matching behavior.
- Added comments to clarify regex flag handling and match count accuracy.
- Adjusted maximum input length comment for clarity.

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
Fix type checks

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
Enhance logging in dataRegexReplaceStep to conditionally display match count based on detailed mode. Add tests for match count behavior with and without global flag. Update approved step definitions hash.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376753)
* Commit: 5794788fbbdaa51b5d8abfe2a0cb6b37a8c3f4ae

### Failed CI Steps
* [FTR Configs #103](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b5498-c36f-4080-9bf7-bc8e824642e3)
* [Scout: [ platform / workflows_extensions ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54b2-f832-4cdf-a84b-693d0d767e14)
* [Scout: [ platform / workflows_extensions ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54c1-19a0-481b-933f-5a10c380e4cb)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54b2-f832-4cdf-a84b-693d0d767e14) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54b2-f832-4cdf-a84b-693d0d767e14/artifacts/019b54c0-d4da-4992-a07b-466756dd7a3e) Scout: [ platform / workflows_extensions ] plugin / serverless-es - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54c1-19a0-481b-933f-5a10c380e4cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54c1-19a0-481b-933f-5a10c380e4cb/artifacts/019b54cd-104d-47f3-a337-d3500e9a21c1) Scout: [ platform / workflows_extensions ] plugin / serverless-es - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54b2-f832-4cdf-a84b-693d0d767e14) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54b2-f832-4cdf-a84b-693d0d767e14/artifacts/019b54c0-d4d9-4550-8df1-d04a94cfed0b) Scout: [ platform / workflows_extensions ] plugin / serverless-oblt - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54c1-19a0-481b-933f-5a10c380e4cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54c1-19a0-481b-933f-5a10c380e4cb/artifacts/019b54cd-104c-4346-9c6e-db30a839d4b5) Scout: [ platform / workflows_extensions ] plugin / serverless-oblt - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54b2-f832-4cdf-a84b-693d0d767e14) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54b2-f832-4cdf-a84b-693d0d767e14/artifacts/019b54c0-d4da-4237-88a1-517a0155df8d) Scout: [ platform / workflows_extensions ] plugin / serverless-security - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54c1-19a0-481b-933f-5a10c380e4cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54c1-19a0-481b-933f-5a10c380e4cb/artifacts/019b54cd-104c-4212-9555-fc5bdf50a0cb) Scout: [ platform / workflows_extensions ] plugin / serverless-security - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54b2-f832-4cdf-a84b-693d0d767e14) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54b2-f832-4cdf-a84b-693d0d767e14/artifacts/019b54c0-d4da-4555-9818-617bccd8ef57) Scout: [ platform / workflows_extensions ] plugin / should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54c1-19a0-481b-933f-5a10c380e4cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54c1-19a0-481b-933f-5a10c380e4cb/artifacts/019b54cd-104d-47a2-8661-608c3469a2ba) Scout: [ platform / workflows_extensions ] plugin / should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54b2-f832-4cdf-a84b-693d0d767e14) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54b2-f832-4cdf-a84b-693d0d767e14/artifacts/019b54c0-d4da-415c-8884-0dc70560afa2) Scout: [ platform / workflows_extensions ] plugin / stateful - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/376753#019b54c1-19a0-481b-933f-5a10c380e4cb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/376753/jobs/019b54c1-19a0-481b-933f-5a10c380e4cb/artifacts/019b54cd-104d-4474-8223-06429b509067) Scout: [ platform / workflows_extensions ] plugin / stateful - Workflows Extensions - Custom Step Definitions Approval - should validate that all registered custom step definitions are approved by workflows-eng team

### History
* :broken_heart: [Build #376694](https://buildkite.com/elastic/kibana-pull-request/builds/376694) failed ec0f2564e41c959e0e255569314ef59d038b07bf
* :broken_heart: [Build #376658](https://buildkite.com/elastic/kibana-pull-request/builds/376658) failed 8b1bdf7d3b04e548a24db0250da92b431a3ee32a
* :broken_heart: [Build #376480](https://buildkite.com/elastic/kibana-pull-request/builds/376480) failed 3b632c193cc8400110e570e2c475c5773cabd40b
* :broken_heart: [Build #376455](https://buildkite.com/elastic/kibana-pull-request/builds/376455) failed 7bb0a11070ab3fa2b13e3ed073fcbf05a17e8be9

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376753","number":376753,"commit":"5794788fbbdaa51b5d8abfe2a0cb6b37a8c3f4ae"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376694","number":376694,"commit":"ec0f2564e41c959e0e255569314ef59d038b07bf"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376658","number":376658,"commit":"8b1bdf7d3b04e548a24db0250da92b431a3ee32a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376480","number":376480,"commit":"3b632c193cc8400110e570e2c475c5773cabd40b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376455","number":376455,"commit":"7bb0a11070ab3fa2b13e3ed073fcbf05a17e8be9"}],"number":376753}
buildkite-pr-comment-->

talboren committed:
Add new approved step definition for data aggregation with corresponding handler hash.

talboren committed:
Update data aggregation step definition and handler hash for improved functionality.

talboren committed:
fix

h88 commented on src/platform/plugins/shared/workflows_extensions/common/steps/data/data_regex_extract_step.ts:
```diff
@@ -0,0 +1,45 @@
+});
+
+export const InputSchema = z.object({
+  pattern: z.string(),
```
consider adding pattern max length validation i.e. z.string().max(10000, 'Pattern exceeds 10K limit')

h88 commented on src/platform/plugins/shared/workflows_extensions/server/steps/data/data_regex_extract_step.ts:
```diff
@@ -0,0 +1,120 @@
+      const extractedItems: Array<Record<string, unknown> | null> = [];
+      let matchCount = 0;
+
+      for (const item of sourceArray) {
```
Noticed in other step impls that we might be able to look for context.abortSignal to allow immediate cancellation (run_stack_monitor.ts:70 // ai_prompt_step.ts:56). Consider adding similar check at the start of each iteration here to prevent processing remaining items after cancellation is requested & ensure workflows can still be cancelled promptly when processing large arrays?

h88 reviewed (commented):
Following up on the previous review consider adding ReDoS pattern pre-validation (maybe in createRegex func) to detect and reject known vulnerable patterns before execution. Common patterns to detect:
- Nested quantifiers: (a+)+, (a*)*, (a+)*
- Overlapping alternations: (a|aa)*, (a|ab|abc)+
- Patterns ending with $ that have nested quantifiers: (a+)+$

This could provide a simple but proactive protection complementing the requested & existing input string and arr len limits in the rest of comments

h88 commented:
<img width="673" height="583" alt="image" src="https://github.com/user-attachments/assets/72953581-9074-46db-bd8f-0b9769128733" />
worth making this console error disappear

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
Add ReDoS pattern detection and validation to regex utilities

- Introduced `detectRedosPatterns` function to identify potentially dangerous regex patterns that could lead to ReDoS vulnerabilities.
- Enhanced `createRegex` function to validate patterns against ReDoS risks before creating RegExp objects.
- Updated tests to cover new detection functionality and ensure safe patterns are allowed.

talboren committed:
Enhance regex pattern validation in data steps

- Updated InputSchema for both data_regex_extract_step and data_regex_replace_step to enforce a maximum pattern length of 10,000 characters.
- Added unit tests to validate the new pattern length restrictions, ensuring that patterns exceeding this limit are rejected while valid patterns are accepted.

talboren committed:
Enhance regex utilities with length validations

- Introduced constants for maximum pattern length (10,000 characters) and maximum array length (10,000 items).
- Updated `createRegex` function to validate pattern length before checking for ReDoS vulnerabilities.
- Enhanced `validateSourceInput` function to reject excessively large arrays, preventing performance issues.
- Added unit tests to ensure proper handling of patterns and arrays at and exceeding their respective length limits.

talboren committed:
abortSignal

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376851)
* Commit: 38a554b45c539c2384fce9acf98022d001bd822f

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/376851#019b63ff-6607-44f0-8221-d05e6d6d8684)

### History
* :broken_heart: [Build #376757](https://buildkite.com/elastic/kibana-pull-request/builds/376757) failed 6f6875fd83743aa438cede165276899504eff610
* :broken_heart: [Build #376753](https://buildkite.com/elastic/kibana-pull-request/builds/376753) failed 5794788fbbdaa51b5d8abfe2a0cb6b37a8c3f4ae
* :broken_heart: [Build #376694](https://buildkite.com/elastic/kibana-pull-request/builds/376694) failed ec0f2564e41c959e0e255569314ef59d038b07bf
* :broken_heart: [Build #376658](https://buildkite.com/elastic/kibana-pull-request/builds/376658) failed 8b1bdf7d3b04e548a24db0250da92b431a3ee32a
* :broken_heart: [Build #376480](https://buildkite.com/elastic/kibana-pull-request/builds/376480) failed 3b632c193cc8400110e570e2c475c5773cabd40b

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376851","number":376851,"commit":"38a554b45c539c2384fce9acf98022d001bd822f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376757","number":376757,"commit":"6f6875fd83743aa438cede165276899504eff610"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376753","number":376753,"commit":"5794788fbbdaa51b5d8abfe2a0cb6b37a8c3f4ae"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376694","number":376694,"commit":"ec0f2564e41c959e0e255569314ef59d038b07bf"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376658","number":376658,"commit":"8b1bdf7d3b04e548a24db0250da92b431a3ee32a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376480","number":376480,"commit":"3b632c193cc8400110e570e2c475c5773cabd40b"}],"number":376851}
buildkite-pr-comment-->

h88 commented on src/platform/plugins/shared/workflows_extensions/server/steps/data/data_regex_extract_step.ts:
```diff
@@ -76,6 +76,14 @@ export const dataRegexExtractStepDefinition = createServerStepDefinition({
 
       for (const item of sourceArray) {
+        // Check for cancellation to allow prompt workflow termination
+        if (context.abortSignal.aborted) {
```
consider ?.

talboren committed:
Fix regex replacement syntax in data_regex_replace_step.ts

- Updated the replacement string to escape dollar signs, ensuring correct pattern substitution.

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into 14805-transform-regex-extract-and-replace

talboren committed:
Enhance regex pattern validation and testing

- Updated `detectRedosPatterns` function to improve detection of dangerous regex patterns, including nested quantifiers with end anchors and overlapping alternations.
- Refactored tests in `data_regex_extract_step.test.ts` and `data_regex_replace_step.test.ts` to use async/await for better handling of asynchronous operations and added validation for error messages related to pattern length.
- Ensured consistent error handling and output validation across regex extraction and replacement steps.

talboren committed:
Update handler hashes in approved step definitions for regex extraction and replacement

- Modified handler hashes for `data.regex_extract` and `data.regex_replace` to reflect recent changes in regex utilities and ensure consistency across the workflow.

jbudz reviewed (approved):
`packages/kbn-optimizer/limits.yml` LGTM

talboren committed:
Merge branch 'main' into 14805-transform-regex-extract-and-replace
